File: #8 Hashing in data structure/hash.py
import random

# Ukuran hash table
TABLE_SIZE = 20

# Membuat hash table kosong
hash_table = [[] for _ in range(TABLE_SIZE)]


# Fungsi hash
def hash_function(key):
    return key % TABLE_SIZE


# INSERT DATA
def insert_data(data):
    index = hash_function(data)

    # Cek agar tidak duplikat
    if data not in hash_table[index]:
        hash_table[index].append(data)
        print(f"Data {data} berhasil ditambahkan.")
    else:
        print("Data sudah ada.")


# HAPUS DATA
def delete_data(data):
    index = hash_function(data)

    if data in hash_table[index]:
        hash_table[index].remove(data)
        print(f"Data {data} berhasil dihapus.")
    else:
        print("Data tidak ditemukan.")


# CARI DATA
def search_data(data):
    index = hash_function(data)

    if data in hash_table[index]:
        print(f"Data {data} ditemukan pada index {index}.")
    else:
        print("Data tidak ditemukan.")


# MENAMPILKAN HASH TABLE
def display_table():
    print("\n=== HASH TABLE ===")
    for i in range(TABLE_SIZE):
        print(f"Index {i}: {hash_table[i]}")


#input 100 angka random
random_numbers = random.sample(range(1, 1000), 100)

for number in random_numbers:
    insert_data(number)

print("\n100 data random berhasil dimasukkan.\n")


# MENU PROGRAM
while True:
    print("\n===== MENU HASH TABLE =====")
    print("1. Input Data")
    print("2. Hapus Data")
    print("3. Cari Data")
    print("4. Tampilkan Hash Table")
    print("5. Keluar")

    choice = input("Pilih menu: ")

    if choice == "1":
        data = int(input("Masukkan angka: "))
        insert_data(data)

    elif choice == "2":
        data = int(input("Masukkan angka yang ingin dihapus: "))
        delete_data(data)

    elif choice == "3":
        data = int(input("Masukkan angka yang ingin dicari: "))
        search_data(data)

    elif choice == "4":
        display_table()

    elif choice == "5":
        print("Program selesai.")
        break

    else:
        print("Pilihan tidak valid.")

File: #8 Hashing in data structure/hash.java
import java.util.*;

public class hash {

    // Ukuran hash table
    static final int TABLE_SIZE = 20;

    // Membuat hash table menggunakan ArrayList
    static LinkedList<Integer>[] hashTable = new LinkedList[TABLE_SIZE];

    // Fungsi hash
    static int hashFunction(int key) {
        return key % TABLE_SIZE;
    }

    // INSERT DATA
    static void insertData(int data) {
        int index = hashFunction(data);

        if (!hashTable[index].contains(data)) {
            hashTable[index].add(data);
            System.out.println("Data " + data + " berhasil ditambahkan.");
        } else {
            System.out.println("Data sudah ada.");
        }
    }

    // HAPUS DATA
    static void deleteData(int data) {
        int index = hashFunction(data);

        if (hashTable[index].contains(data)) {
            hashTable[index].remove(Integer.valueOf(data));
            System.out.println("Data " + data + " berhasil dihapus.");
        } else {
            System.out.println("Data tidak ditemukan.");
        }
    }

    // CARI DATA
    static void searchData(int data) {
        int index = hashFunction(data);

        if (hashTable[index].contains(data)) {
            System.out.println("Data " + data + " ditemukan pada index " + index);
        } else {
            System.out.println("Data tidak ditemukan.");
        }
    }

    // TAMPILKAN HASH TABLE
    static void displayTable() {
        System.out.println("\n=== HASH TABLE ===");

        for (int i = 0; i < TABLE_SIZE; i++) {
            System.out.print("Index " + i + ": ");

            for (Integer data : hashTable[i]) {
                System.out.print(data + " ");
            }

            System.out.println();
        }
    }

    public static void main(String[] args) {

        // Inisialisasi setiap index
        for (int i = 0; i < TABLE_SIZE; i++) {
            hashTable[i] = new LinkedList<>();
        }

        // Generate 100 angka random unik
        Random random = new Random();
        HashSet<Integer> uniqueNumbers = new HashSet<>();

        while (uniqueNumbers.size() < 100) {
            int number = random.nextInt(1000) + 1;
            uniqueNumbers.add(number);
        }

        // Masukkan ke hash table
        for (int number : uniqueNumbers) {
            insertData(number);
        }

        System.out.println("\n100 data random berhasil dimasukkan.");

        // Menu program
        Scanner input = new Scanner(System.in);

        while (true) {
            System.out.println("\n===== MENU HASH TABLE =====");
            System.out.println("1. Input Data");
            System.out.println("2. Hapus Data");
            System.out.println("3. Cari Data");
            System.out.println("4. Tampilkan Hash Table");
            System.out.println("5. Keluar");

            System.out.print("Pilih menu: ");
            int choice = input.nextInt();

            switch (choice) {

                case 1:
                    System.out.print("Masukkan angka: ");
                    int insert = input.nextInt();
                    insertData(insert);
                    break;

                case 2:
                    System.out.print("Masukkan angka yang ingin dihapus: ");
                    int delete = input.nextInt();
                    deleteData(delete);
                    break;

                case 3:
                    System.out.print("Masukkan angka yang ingin dicari: ");
                    int search = input.nextInt();
                    searchData(search);
                    break;

                case 4:
                    displayTable();
                    break;

                case 5:
                    System.out.println("Program selesai.");
                    System.exit(0);

                default:
                    System.out.println("Pilihan tidak valid.");
            }
        }
    }
}